settle-bot/src/settle.rs

```diff
@@ -0,0 +1,315 @@
+                    } else {
+                        liq_end_health
+                    };
+
+                    let fee = perp_market
+                        .compute_settle_fee(settleable, liq_end_health, maint_health)
+                        .unwrap();
+                    if fee <= 0 {
+                        continue;
+                    }
+
+                    all_positive_settleable
+                        .entry(perp_market_index)
+                        .or_default()
+                        .push((*account_key, settleable, fee));
+                } else if settleable < 0 {
+                    all_negative_settleable
+                        .entry(perp_market_index)
+                        .or_default()
+                        .push(*account_key, -settleable);
+                }
+            }
+        }
+
+        let address_lookup_tables = mango_client.mango_address_lookup_tables().await?;
+
+        for (perp_market_index, mut positive_settleable) in all_positive_settleable {
+            let (perp_market, _) = perp_market_info.get(&perp_market_index).unwrap();
+            let negative_settleable = match all_negative_settleable.get_mut(&perp_market_index) {
+                None => continue,
+                Some(v) => v,
+            };
+            // sort by fee, descending
+            positive_settleable.sort_by_key(|v| -v.2);
+
+            let mut batch_processor = SettleBatchProcessor {
+                mango_client,
+                account_fetcher,
+                perp_market_index,
+                instructions: Vec::new(),
+                max_batch_size: 8, // the 1.4M max CU limit if we assume settle ix can be up to around 150k
+                blockhash: mango_client
+                    .client
+                    .rpc_async()
+                    .get_latest_blockhash_with_commitment(CommitmentConfig::confirmed())
+                    .await?
+                    .0,
+                address_lookup_tables: &address_lookup_tables,
+            };
+            for (account_a, settleable_a, fee) in positive_settleable {
+                // find the best remaining counterparty
+                let (&account_b, &settleable_b) = match negative_settleable.peek() {
+                    None => break,
+                    Some(v) => v,
+                };
+
+                let settleable = settleable_a.min(settleable_b);
+                if settleable <= 0
+                    || (settleable < perp_market.settle_fee_amount_threshold
+                        && fee == perp_market.settle_fee_flat)
+                {
+                    // no more interesting pairs that would produce fees
+                    break;
+                }
+
+                // add to batch and maybe send
+                batch_processor
+                    .add_and_maybe_send(account_a, account_b)
+                    .await?;
+                negative_settleable.change_priority(&account_b, settleable_b - settleable);
```

Comment:
hmm do we keep on reusing account_a , settleable_a, without shaving off settleable_a after we settle..similar to how we do for account_b?

Reply:
I'll add a note about that. We loop through the positive-pnl accounts exactly once, so modifying these values doesn't matter.

Reply:
ah, but you could have a mm, who has a large position and be +ve pnl, means the distribution won't be smooth right

Reply:
Yeah, if there's an account with lots of +pnl and all accounts with -pnl are small then that could take multiple passes to get settled right now. I'll add that to the note.

Reply:
actually did a nice solution now